pkg/logql/first_over_time.go

```diff
@@ -0,0 +1,189 @@
+package logql
+
+import (
+	"math"
+	"time"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql"
+
+	"github.com/grafana/loki/pkg/iter"
+)
+
+func newFirstWithTimestampIterator(
+	it iter.PeekingSampleIterator,
+	selRange, step, start, end, offset int64) RangeVectorIterator {
+	inner := &batchRangeVectorIterator{
+		iter:     it,
+		step:     step,
+		end:      end,
+		selRange: selRange,
+		metrics:  map[string]labels.Labels{},
+		window:   map[string]*promql.Series{},
+		agg:      nil,
+		current:  start - step, // first loop iteration will set it to start
+		offset:   offset,
+	}
+	return &firstWithTimestampBatchRangeVectorIterator{
+		batchRangeVectorIterator: inner,
+	}
+}
+
+type firstWithTimestampBatchRangeVectorIterator struct {
+	*batchRangeVectorIterator
+	at []promql.Sample
+}
+
+func (r *firstWithTimestampBatchRangeVectorIterator) At() (int64, StepResult) {
+	if r.at == nil {
+		r.at = make([]promql.Sample, 0, len(r.window))
+	}
+	r.at = r.at[:0]
+	// convert ts from nano to milli seconds as the iterator work with nanoseconds
+	ts := r.current/1e+6 + r.offset/1e+6
+	for _, series := range r.window {
+		s := r.agg(series.Floats)
+		r.at = append(r.at, promql.Sample{
+			F:      s.F,
+			T:      s.T / int64(time.Millisecond),
```

Comment:
Vlad mentioned we should be using nanos here. What do you think?

Reply:
I don't think it matters? When we return the result to users we interploate forwards in time to the next interval of the `Step` anyways. If we were going to change it, we should probably not use internal prometheus types, since prometheus is only ever millisecond precise. It might not be a problem right now, but we shouldn't rely on something _not_ changing in a way that would make that millisecond precision more strict in some way.